benchmark/src/steps.ts

```diff
@@ -0,0 +1,265 @@
+        },
+        false
+    ),
+    new BenchmarkStep('newImmerDraft', ({ freezeState }) => {
+        const { actionHistory, initialState } = freezeState!;
+
+        // Apply each action on a new immer draft
+        return actionHistory.reduce(
+            (state, action) => reduceExerciseState(state, action),
+            initialState
+        );
+    }),
+    new BenchmarkStep('sameImmerDraft', ({ freezeState }) => {
+        const { actionHistory, initialState } = freezeState!;
+
+        // Apply all action on the same immer draft
+        return produce(initialState, (draftState) => {
+            for (const action of actionHistory) {
+                applyAction(draftState, action);
+            }
+        });
+    }),
+    new BenchmarkStep('noImmerDraft', ({ freezeState }) => {
+        const { actionHistory, initialState } = freezeState!;
+
+        // Apply all action on the same immer draft
+        return actionHistory.reduce(
+            (state, action) => applyAction(state, action),
+            cloneDeepMutable(initialState)
+        );
+    }),
+    new CalculationStep(
+        'endStatesAreEqual',
+        ({ newImmerDraft, sameImmerDraft, noImmerDraft }) => {
+            const endStatesAreEqual =
+                isEqual(newImmerDraft!.value, sameImmerDraft!.value) &&
+                isEqual(newImmerDraft!.value, noImmerDraft!.value);
+            if (!endStatesAreEqual) {
+                print(
+                    `  The endStates of the previous three steps are not equal!
+  This most likely means that a reducer is either not deterministic or makes some assumptions about immer specific stuff (use of "original()").
+  To further debug this, you should log the endStates of the respective exercises and can compare them directly in vscode via "Compare file with".
+`,
+                    'red'
+                );
+            }
+            return endStatesAreEqual;
+        },
+        false
+    ),
+
+    new CalculationStep(
+        'benchmarkActions',
+        ({ freezeState }) => {
+            print(`  benchmarkActions: `);
+            const { actionHistory, initialState } = freezeState!;
+            const totalTimePerAction: {
+                [key in ExerciseAction['type']]?: number;
+            } = {};
+            let currentState = initialState;
+            for (const action of actionHistory) {
+                // eslint-disable-next-line @typescript-eslint/no-loop-func
+                const { value: newState, time } = benchmark(() =>
+                    reduceExerciseState(currentState, action)
+                );
+
+                currentState = newState;
+                totalTimePerAction[action.type] =
+                    (totalTimePerAction[action.type] ?? 0) + time;
+            }
+            const sortedTotalTimePerAction = sortObject(
+                totalTimePerAction,
+                ([, timeA], [, timeB]) => timeB! - timeA!
+            );
+            print(
```

Comment:
I'm also wondering whether these outputs (both here and for `numberOfActionsPerType` below) are well formatted. For a large exercise, this is just a huge wall of text. I can't think of a better way currently, though, so no concrete suggestion here (for now).

Reply:
I don't think this needs to be very pretty. If you want to check whether a specific action was in there, you can do so and else it is still relatively easy to skim the rest of the output.